circuits/cpp/barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/eccvm/msm_builder.hpp

```diff
@@ -0,0 +1,263 @@
+            } else {
+                // 8 maps to 7
+                // 15 maps to 0
+
+                // 15 - x
+                point_table_read_counts[column_index][pc_offset + 15 - static_cast<size_t>(slice_row)]++;
+            }
+            // slice : row
+            // -15 : 0
+            // -13 : 1
+            // -11 : 2
+            // -9 : 3
+            // -7 : 4
+            // -5 : 5
+            // -3 : 6
+            // -1 : 7
+            // 1 : 8
+            // 3 : 9
+            // 5 : 10
+            // 7 : 11
+            // 9 : 12
+            // 11 : 13
+            // 13 : 14
+            // 15 : 15
+        };
+        std::vector<MSMState> msm_state;
+        // start with empty row (shiftable polynomials must have 0 as first coefficient)
+        msm_state.emplace_back(MSMState{});
+        uint32_t pc = total_number_of_muls;
+        AffineElement accumulator = CycleGroup::affine_point_at_infinity;
+
+        for (const auto& msm : msms) {
+            const size_t msm_size = msm.size();
+
+            const size_t rows_per_round = (msm_size / ADDITIONS_PER_ROW) + (msm_size % ADDITIONS_PER_ROW != 0 ? 1 : 0);
+            static constexpr size_t num_rounds = NUM_SCALAR_BITS / WNAF_SLICE_BITS;
+
+            const auto add_points = [](auto& P1, auto& P2, auto& lambda, auto& collision_inverse, bool predicate) {
+                lambda = predicate ? (P2.y - P1.y) / (P2.x - P1.x) : 0;
+                collision_inverse = predicate ? (P2.x - P1.x).invert() : 0;
+                auto x3 = predicate ? lambda * lambda - (P2.x + P1.x) : P1.x;
+                auto y3 = predicate ? lambda * (P1.x - x3) - P1.y : P1.y;
+                return AffineElement(x3, y3);
+            };
+            for (size_t j = 0; j < num_rounds; ++j) {
+                for (size_t k = 0; k < rows_per_round; ++k) {
+                    MSMState row;
+                    const size_t points_per_row =
+                        (k + 1) * ADDITIONS_PER_ROW > msm_size ? msm_size % ADDITIONS_PER_ROW : ADDITIONS_PER_ROW;
+                    const size_t idx = k * ADDITIONS_PER_ROW;
+                    row.msm_transition = (j == 0) && (k == 0);
+
+                    AffineElement acc(accumulator);
+                    Element acc_expected = accumulator;
+                    for (size_t m = 0; m < 4; ++m) {
+                        auto& add_state = row.add_state[m];
+                        add_state.add = points_per_row > m;
+                        int slice = add_state.add ? msm[idx + m].wnaf_slices[j] : 0;
+                        add_state.slice = add_state.add ? (slice + 15) / 2 : 0;
+                        add_state.point = add_state.add
+                                              ? msm[idx + m].precomputed_table[static_cast<size_t>(add_state.slice)]
+                                              : AffineElement{ 0, 0 };
+                        bool add_predicate = (m == 0 ? (j != 0 || k != 0) : add_state.add);
```

Comment:
A comment here would help understand the logic of the predicate

Reply:
added comment

Reply:
oh I commented on the predicate 3 lines above, will fix and add comments for this one too